lib/spack/spack/stage.py

```
@@ -725,6 +738,148 @@ def _add_to_root_stage(self):
+    to perform as normal.
+    The external build tree is given the same lifespan as its stage dir, is spun up when the stage
+    spins up, and destroyed when the stage is destroyed. After the build and installation are done,
+    the build tree is first relocated to what would be its proper place in the stage.
+    and a symlink is placed in place of the build tree pointing at the external directory
+
+    Note: This is not, nor should it be, used on *nix platforms and is intended as a solution to
+    reduce file path lengths on Windows during compilation/linking. This class should be removed
+    when MSVC fully supports the LongPath feature on Windows.
+    """
+
+
+    def __init__(self, hash, fetcher, **kwargs):
+        # Users can override external cmake build dir, default is %USERPROFILE%
+        # overrides can come from command line or config, command line will override all
+        super(CMakeBuildStage, self).__init__(fetcher, **kwargs)
+        self._hash = hash
+        self._remote_stage = None
+        self._remote_root = Path(spack.config.get("config:cmake_ext_build"))
+
+    @property
+    def root_stage_context(self):
+        return self._path / ("spack-build-%s" % self._hash)
+
+    @property
+    def build_directory(self):
+        return self._remote_stage
+
+    def _establish_context_link(self):
+        symlink(str(self._remote_stage), str(self.root_stage_context))
+
+    def _remove_context_link(self):
+        self.root_stage_context.unlink()
+
+    def _setup_remote_build_stage(self):
+        # try to create root if it doesn't exist
+        self._remote_root.mkdir(parents=True, exist_ok=True)
+        sub_dir = self._compute_next_open_subdir()
+        full_subdir = self._remote_root / sub_dir
+        attempts = 0
+        while True and attempts < 200:
+            try:
+                full_subdir.mkdir()
+                break
+            except FileExistsError:
```

I recommend removing the try/except (and the loop that re-attempts creating the file)

* You would need to put `_compute_next_open_subdir()` in this loop, or else it will fail 200 times
* This would need to raise an exception or else it will end up just returning the failed directory after reaching the max number of attempts (i.e. someone else's directory)
* Since `_compute_next_open_subdir` attempts to account for already-existing directories, the only time this exception would be triggered is in a multi-process environment. AFAIK `.mkdir` is not atomic though, so we'd have to be more-careful in this case anyway. It might be easier overall to assume serial and skip the try/except altogether (also we can assume that whatever root directory we are using, spack is the only one using it).